contextualized/dags/torch_notmad/torch_notmad.py

```diff
@@ -0,0 +1,190 @@
+        explainer_output_shape = (self.feature_shape[1], self.feature_shape[1])
+
+        # layer params
+        self.init_mat = init_mat
+
+        # opt params
+        self.learning_rate = learning_rate
+        self.opt_step = opt_step
+
+        # layers
+        self.encoder = NGAM(
+            encoder_input_shape[0],
+            encoder_output_shape[0],
+            layers=encoder_kwargs["layers"],
+            width=encoder_kwargs["width"],
+        )
+        self.explainer = Explainer(self.n_archetypes, explainer_output_shape)
+
+        # loss
+        self.my_loss = lambda x, y: self._build_arch_loss(
+            archetype_loss_params
+        ) + NOTEARS_loss(x, y, sample_specific_loss_params["l1"], self.alpha, self.rho)
+
+    def forward(self, c):
+        c = c.float()
+        c = self.encoder(c).float()
+        out = self.explainer(c)
+        return out.float()
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+        sch = torch.optim.lr_scheduler.StepLR(
+            optimizer, step_size=self.opt_step, gamma=0.5
+        )
+        # learning rate scheduler
+        return {
+            "optimizer": optimizer,
+            "lr_scheduler": {
+                "scheduler": sch,
+                "monitor": "train_loss",
+            },
+        }
+
+    def training_step(self, batch, batch_idx):
+        C, x_true = batch
+        w_pred = self.forward(C).float()
+        loss = self.my_loss(x_true.float(), self._mask(w_pred)).float()
+        self.log("train_loss", loss)
+
+        return loss
+
+    def test_step(self, batch, batch_idx):
+        C, x_true = batch
+        w_pred = self.forward(C).float()
+        loss = self.my_loss(x_true.float(), w_pred.float())
+        self.log("test_loss", loss)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        C, x_true = batch
+        w_pred = self.forward(C).float()
+        # print(w_pred[0])
+        loss = self.my_loss(x_true.float(), w_pred.float())
+        self.log("val_loss", loss)
+        return loss
+
+    def predict_step(self, batch, batch_idx):
+        C, x_true = batch
+        w_pred = self.forward(C.float()).float()
+        return w_pred
+
+    def predict_w(self, C, confirm_project_to_dag=False):
+        w_preds = self.forward(torch.tensor(C))
+
+        if confirm_project_to_dag:
+            try:
+                return np.array(
+                    [project_to_dag_torch(w.detach().numpy())[0] for w in w_preds]
+                )
+            except:
+                print("Error, couldn't project to dag. Returning normal predictions.")
+
+        return w_preds
+
+    # helpers
+    def _mask(self, W):
+        d = W.shape[1]
+        mask = torch.ones(d).float() - torch.eye(d).float()
+        return torch.multiply(W, mask)
+
+    def _parse_alpha_rho(self, params):
+        alpha = params["alpha"]
+        rho = params["rho"]
+        dynamic = False
```

Comment:
Should this be a parameter?

Reply:
It looks like DynamicAlphaRho has to be specified outside NOTAMD in the Trainer right now. We should probably make this internal to NOTMAD if possible instead of spreading it over two objects.

Reply:
Looked into this and was actually just able to get dynamic alpha rho to now work as not a callback (in most recent commit).